pandas/core/arrays/categorical.py

```diff
@@ -402,20 +402,40 @@ def astype(self, dtype: Dtype, copy: bool = True) -> ArrayLike:
             If copy is set to False and dtype is categorical, the original
             object is returned.
         """
-        if is_categorical_dtype(dtype):
+        if self.dtype is dtype:
+            result = self.copy() if copy else self
+
+        elif is_categorical_dtype(dtype):
             dtype = cast(Union[str, CategoricalDtype], dtype)
 
             # GH 10696/18593/18630
             dtype = self.dtype.update_dtype(dtype)
-            result = self.copy() if copy else self
-            if dtype == self.dtype:
-                return result
-            return result._set_dtype(dtype)
-        if is_extension_array_dtype(dtype):
-            return array(self, dtype=dtype, copy=copy)
-        if is_integer_dtype(dtype) and self.isna().any():
+            self = self.copy() if copy else self
+            result = self._set_dtype(dtype)
+
+        # TODO: consolidate with ndarray case?
+        elif is_extension_array_dtype(dtype):
+            result = array(self, dtype=dtype, copy=copy)
+
+        elif is_integer_dtype(dtype) and self.isna().any():
             raise ValueError("Cannot convert float NaN to integer")
-        return np.array(self, dtype=dtype, copy=copy)
+
+        elif len(self.codes) == 0 or len(self.categories) == 0:
+            result = np.array(self, dtype=dtype, copy=copy)
+
+        else:
+            # GH8628 (PERF): astype category codes instead of astyping array
+            try:
+                astyped_cats = self.categories.astype(dtype=dtype, copy=copy)
+            except (TypeError, ValueError):
+                raise ValueError(
```

Comment:
why change TypeError?

Reply:
It's to fix the error message for `CategoricalIndex`. If we don't catch `TypeError` we end up with `TypeError: Cannot cast Index to dtype float64` (below) versus something like `TypeError: Cannot cast object to dtype float64`
``` python
In [2]: idx = pd.CategoricalIndex(["a", "b", "c", "a", "b", "c"])

In [3]: idx.astype('float')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/workspaces/pandas-arw2019/pandas/core/indexes/base.py in astype(self, dtype, copy)
    700         try:
--> 701             casted = self._values.astype(dtype, copy=copy)
    702         except (TypeError, ValueError) as err:

ValueError: could not convert string to float: 'a'

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
<ipython-input-4-38d56ec15c36> in <module>
----> 1 idx.astype('float')

/workspaces/pandas-arw2019/pandas/core/indexes/category.py in astype(self, dtype, copy)
    369     @doc(Index.astype)
    370     def astype(self, dtype, copy=True):
--> 371         res_data = self._data.astype(dtype, copy=copy)
    372         return Index(res_data, name=self.name)
    373 

/workspaces/pandas-arw2019/pandas/core/arrays/categorical.py in astype(self, dtype, copy)
    427             # GH8628 (PERF): astype category codes instead of astyping array
    428             try:
--> 429                 astyped_cats = self.categories.astype(dtype=dtype, copy=copy)
    430             except (ValueError):
    431                 raise ValueError(

/workspaces/pandas-arw2019/pandas/core/indexes/base.py in astype(self, dtype, copy)
    701             casted = self._values.astype(dtype, copy=copy)
    702         except (TypeError, ValueError) as err:
--> 703             raise TypeError(
    704                 f"Cannot cast {type(self).__name__} to dtype {dtype}"
    705             ) from err

TypeError: Cannot cast Index to dtype float64
```

Reply:
ok this is fine, but can you add a comment for this then right here (so future readers understand)